src/cloud-api-adaptor/docker/README.md

```diff
@@ -80,9 +78,68 @@ For changing the CAA image to your custom built image (eg. `quay.io/myuser/cloud
 you can use the following:
 
 ```bash
-kubectl set image ds/cloud-api-adaptor-daemonset -n confidential-containers-system cloud-api-adaptor-con=quay.io/myuser/cloud-api-adaptor
+export CAA_IMAGE=quay.io/myuser/cloud-api-adaptor
+kubectl set image ds/cloud-api-adaptor-daemonset -n confidential-containers-system cloud-api-adaptor-con="$CAA_IMAGE"
 ```
 
+## Running the CAA e2e tests
+
+### Test Prerequisites
+
+Following prerequisites are needed on the test system:
+
+- make
+- go
+- yq
+- kubectl
+- kind
+- docker
+  
+A `prereqs.sh` helper script is available under `src/cloud-api-adaptor/docker` to install/uninstall the prerequisites.
+
+Note: If using the `prereqs.sh` helper script to install the prerequisites, then reload the shell to ensure new permissions
+are in place to run `docker` and other commands.
+
+### Test Execution
+
+In order to run the tests, edit the file `src/cloud-api-adaptor/test/provisioner/docker/provision_docker.properties`
+and update the `CAA_IMAGE` and `CAA_IMAGE_TAG` variables with your custom CAA image and tag.
+
+You can run the CAA e2e [tests/e2e/README.md](../test/e2e/README.md) by running the following command:
+
+```sh
+TEST_PODVM_IMAGE=<podvm-image> TEST_PROVISION=yes CLOUD_PROVIDER=docker TEST_PROVISION_FILE=$(pwd)/test/provisioner/docker/provision_docker.properties test-e2e
+```
+
+This will create a two node kind cluster, automatically download the pod VM image mentioned in the `provision_docker.properties`
+file and run the tests.
+
+Note: To overcome docker rate limiting issue or to download images from private registries,
+create a `config.json` file under `/tmp` with your registry secrets.
+For example:
+If your docker registry user is `someuser` and password is `somepassword` then create the auth string
+as shown below:
+
+```sh
+echo -n "someuser:somepassword" | base64
+c29tZXVzZXI6c29tZXBhc3N3b3Jk
+```
+
+This auth string needs to be used in `/tmp/config.json` as shown below:
+
+```sh
+{
+        "auths": {
+                "https://index.docker.io/v1/": {
+                        "auth": "c29tZXVzZXI6c29tZXBhc3N3b3Jk"
+                }
+        }
+}
+```
+
+If you want to use a different location for the registry secret, then remember to update the same
+in the `docker/kind-config.yaml` file.
+
```

Comment:
The cluster will be deleted automatically on test completion unless `TEST_TEARDOWN=no`. Anyways let me add a note as well.

Reply:
So that might not have worked for me then as I copied your command from the PR, but my cluster is over two hours old now from all the stopped and re-testing. I thought that I'd let the e2e process finish naturally at least once, but maybe not. It might be better to note that the e2e test using kind to create the cluster and just link to their getting start docs instead?

Reply:
I think this was user error. I'm just re-trying and found that the `Uninstall CCRuntime CRD` step has taken 5mins so far, so I probably killed it previously thinking it was stuck/finished

Reply:
Hmm, maybe not - it failed badly, so didn't do the kind delete:
```
time="2024-07-01T10:10:08-07:00" level=info msg="Delete the peerpod-ctrl deployment"
FAIL	github.com/confidential-containers/cloud-api-adaptor/src/cloud-api-adaptor/test/e2e	391.579s
FAIL
make: *** [Makefile:96: test-e2e] Error 1
```
I will have some more attempts to see if I can track down the issue

Reply:
Same here @stevenhorsman , the kind cluster was left behind. Maybe this is a bug on the test framework itself that if something goes wrong on teardown then it won't run until the end (and the last step is to delete the cluster?)

```
time="2024-07-01T19:43:00Z" level=info msg="Deleting namespace 'coco-pp-e2e-test-5a79bbe1'..."
time="2024-07-01T19:43:15Z" level=info msg="Namespace 'coco-pp-e2e-test-5a79bbe1' has been successfully deleted within 60s"
Deleting the kind cluster
Deleting cluster "kind" ...
time="2024-07-01T19:43:15Z" level=info msg="Delete the Cloud API Adaptor installation"
time="2024-07-01T19:43:15Z" level=info msg="Uninstall the cloud-api-adaptor"
time="2024-07-01T19:43:15Z" level=info msg="Uninstall CCRuntime CRD"
time="2024-07-01T19:47:27Z" level=info msg="Uninstall the controller manager"
time="2024-07-01T19:47:36Z" level=info msg="Wait for the cc-operator-controller-manager deployment be deleted\n"
time="2024-07-01T19:47:41Z" level=info msg="Delete the peerpod-ctrl deployment"
FAIL    github.com/confidential-containers/cloud-api-adaptor/src/cloud-api-adaptor/test/e2e     1316.300s
FAIL
make: *** [Makefile:98: test-e2e] Error 1
```

Reply:
Yeah,

FYI with trace on I got:
```
time="2024-07-02T01:37:00-07:00" level=trace msg="/usr/bin/make -C ../peerpod-ctrl undeploy, output: make[1]: Entering directory '/root/go/src/github.com/confidential-containers/cloud-api-adaptor/src/peerpod-ctrl'\n/root/go/src/github.com/confidential-containers/cloud-api-adaptor/src/peerpod-ctrl/bin/kustomize build config/default | kubectl delete --ignore-not-found=false -f -\n# Warning: 'patchesStrategicMerge' is deprecated. Please use 'patches' instead. Run 'kustomize edit fix' to update your Kustomization automatically.\ncustomresourcedefinition.apiextensions.k8s.io \"peerpods.confidentialcontainers.org\" deleted\nclusterrole.rbac.authorization.k8s.io \"peerpod-ctrl-manager-role\" deleted\nclusterrole.rbac.authorization.k8s.io \"peerpod-ctrl-metrics-reader\" deleted\nclusterrole.rbac.authorization.k8s.io \"peerpod-ctrl-proxy-role\" deleted\nclusterrolebinding.rbac.authorization.k8s.io \"peerpod-ctrl-manager-rolebinding\" deleted\nclusterrolebinding.rbac.authorization.k8s.io \"peerpod-ctrl-proxy-rolebinding\" deleted\nError from server (NotFound): error when deleting \"STDIN\": namespaces \"confidential-containers-system\" not found\nError from server (NotFound): error when deleting \"STDIN\": serviceaccounts \"peerpod-ctrl-controller-manager\" not found\nError from server (NotFound): error when deleting \"STDIN\": roles.rbac.authorization.k8s.io \"peerpod-ctrl-leader-election-role\" not found\nError from server (NotFound): error when deleting \"STDIN\": rolebindings.rbac.authorization.k8s.io \"peerpod-ctrl-leader-election-rolebinding\" not found\nError from server (NotFound): error when deleting \"STDIN\": services \"peerpod-ctrl-controller-manager-metrics-service\" not found\nError from server (NotFound): error when deleting \"STDIN\": deployments.apps \"peerpod-ctrl-controller-manager\" not found\nmake[1]: *** [Makefile:182: undeploy] Error 1\nmake[1]: Leaving directory '/root/go/src/github.com/confidential-containers/cloud-api-adaptor/src/peerpod-ctrl'\n"
```

Reply:
Raised https://github.com/confidential-containers/cloud-api-adaptor/issues/1898 for this